packages/sdk/schema/src/testing/generator.ts

```diff
@@ -0,0 +1,162 @@
+// TODO(burdon): Generate views, tables, and SCRIPTS/FUNCTIONS.
+// TODO(burdon): Generate test documents, sketches, sheets.
+
+/**
+ * Set properties based on generator annotation.
+ */
+export const setProps = <T extends BaseObject>(type: S.Schema<T>) => {
+  return (initial: ExcludeId<T> = {} as ExcludeId<T>): Effect.Effect<ExcludeId<T>> => {
+    const obj = getSchemaProperties<T>(type.ast).reduce<ExcludeId<T>>((data, property) => {
+      if (data[property.name] === undefined) {
+        const gen = findAnnotation<string>(property.ast, GeneratorAnnotationId);
+        if (gen) {
+          const fn = getDeep<() => any>({ faker }, gen.split('.'));
+          if (fn) {
+            data[property.name] = fn() as any;
+          }
+        } else if (!property.optional) {
+          log.warn('missing generator for required property', { property });
+        }
+      }
+
+      return data;
+    }, initial);
+
+    return Effect.succeed(obj);
+  };
+};
+
+/**
+ * Set references.
+ */
+export const setReferences = <T extends BaseObject>(type: S.Schema<T>, db: EchoDatabase) => {
+  return (obj: BaseObject<T>) => {
+    return Effect.promise(async () => {
+      for (const property of getSchemaProperties<T>(type.ast)) {
+        if (property.format === FormatEnum.Ref) {
+          const jsonSchema = findAnnotation<JsonSchemaType>(property.ast, AST.JSONSchemaAnnotationId);
+          if (jsonSchema) {
+            const typename = getSchemaReference(jsonSchema);
+            invariant(typename);
+            // TODO(burdon): Filter.typename doesn't work!
+            const { objects } = await db.query((obj) => getTypename(obj) === typename).run();
+            if (objects.length && faker.datatype.boolean()) {
+              const object = faker.helpers.arrayElement(objects);
+              obj[property.name] = object;
+            }
+          }
+        }
+      }
+
+      return obj;
+    });
+  };
+};
+
+export const createReactiveObject = <T extends BaseObject<any>>(type: S.Schema<T>) => {
+  return (data: ExcludeId<T>): Effect.Effect<BaseObject<T>> => {
+    return Effect.succeed(create<T>(type, data));
+  };
+};
+
+export const addToDatabase = (db: EchoDatabase) => {
+  return <T extends BaseObject<any>>(obj: ReactiveObject<T>): Effect.Effect<BaseObject<T>> => {
+    return Effect.succeed(db.add(obj));
+  };
+};
+
+//
+// Effect pipeline.
+// - Allows for mix of sync and async transformations.
+// - Consistent error processing.
+// TODO(burdon): Pass db in context?
+//
+
+export const noop = (obj: any) => obj;
+
+export const logObject = (message: string) => (obj: any) => log.info(message, { obj });
+
+export const createObjectArray = <T extends BaseObject<T>>(n: number): ExcludeId<T>[] =>
+  Array.from({ length: n }, () => ({}) as any);
+
+export const createArrayPipeline = <T extends BaseObject<any>>(
+  n: number,
+  pipeline: (obj: ExcludeId<T>) => Effect.Effect<ExcludeId<T>>,
```

Comment:
What do you think about having a non-effect API for this so it's consistent with the rest of the codebase?